### PR TITLE
Feature/giabeval_multiqc

### PR DESCRIPTION
**Pull request by melferink**
* Feature GIABEval
* code from https://github.com/UMCUGenetics/CustomModules/tree/feature/giabeval re-commited for minimal changes needed for GIABEval

### CODE REVIEW

File: CheckQC/CheckQC.nf [hunk 1]
```diff
@@ -7,7 +7,7 @@ process CheckQC {
     input:
         val(identifier)
         path(input_files)
-     
+
     output:
         path("${identifier}_summary.csv", emit: qc_output)
```
Comment (Jorisvansteenbrugge) [inline CheckQC/CheckQC.nf:12]: Blijkbaar werkte dit, maar het is netter om emit buiten de `path()` te hebben staan
```suggestion
        path("${identifier}_summary.csv"), emit: qc_output
```

File: CheckQC/check_qc.py [hunk 2]
```diff
@@ -176,8 +176,20 @@ def create_and_write_output(qc_output, output_path, output_prefix):
...
+    """
+    Read and judge each single qc metric and join results.
+
+    Args:
+        qc (dict): qc settings of the metric
+        metrics (list): List of input files specific for single qc metric
+
+    Returns:
+        pandas DataFrame: Joined and judged qc metrics.
+    """
```
Comment (Jorisvansteenbrugge) [inline CheckQC/check_qc.py:188]: As someone that has never seen this function before, this docstring seems incomplete. 
1. What are `QC settings of the metric`? 
2. Metrics is a list of QC files, but is that a QC file for one sample, or are there multiple QC files per sample?
3. What is a joined and judged qc metrics? Maybe you can describe the layout a little bit

File: CheckQC/check_qc.py [hunk 3]
```diff
@@ -196,6 +208,8 @@ def read_and_judge_metrics(qc, metrics):
             # Check for duplicate sampleIDs before merge.
             if any(qc_judged_renamed["sample"].isin(output["sample"])):
                 is_duplicate_sample = True
+            output = output.astype(object)
+            qc_judged_renamed = qc_judged_renamed.astype(object)
```
Comment (Jorisvansteenbrugge) [inline CheckQC/check_qc.py:212]: Some comments here would be nice to explain what is happening. It seems like you are casting the output(a pandas dataframe) to an `object` type, but what doest that imply/why is it necessary?

File: Utils/EditSummaryFileHappy.nf [hunk 4]
```diff
@@ -0,0 +1,30 @@
+process EditSummaryFileHappy {
+    tag "$meta.id"
+    label 'EditSummaryFileHappy'
+    shell = ['/bin/bash', '-euo', 'pipefail']
+
+    input:
+        // meta should have the keys 'id', 'query' and 'truth'
+        tuple(val(meta), path(summary_csv))
```
Comment (Jorisvansteenbrugge) [inline Utils/EditSummaryFileHappy.nf:8]:
```suggestion
        tuple val(meta), path(summary_csv)
```
To improve readability

File: Utils/EditSummaryFileHappy.nf [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    output:
+        path("${meta.truth}_${meta.query}_INDEL_PASS.summary.csv"), emit: indel_pass_csv
+        path("${meta.truth}_${meta.query}_INDEL_ALL.summary.csv"), emit: indel_all_csv
+        path("${meta.truth}_${meta.query}_SNP_PASS.summary.csv"), emit: snp_pass_csv
+        path("${meta.truth}_${meta.query}_SNP_ALL.summary.csv"), emit: snp_all_csv
```
Comment (Jorisvansteenbrugge) [inline Utils/EditSummaryFileHappy.nf:14]: For future reference, you don't have to specify the complete output paths, this is slightly more flexible if you would change any inputs later on. 
```suggestion
        path("*_INDEL_PASS.summary.csv"), emit: indel_pass_csv
        path("*_INDEL_ALL.summary.csv"), emit: indel_all_csv
        path("*_SNP_PASS.summary.csv"), emit: snp_pass_csv
        path("*_SNP_ALL.summary.csv"), emit: snp_all_csv
```

**Review by Jorisvansteenbrugge — CHANGES_REQUESTED** [review]
Where are the unit tests?  👀

File: CheckQC/CheckQC.nf (hunk 1 above)
Comment (melferink) [reply CheckQC/CheckQC.nf:12]: Fixed

File: Utils/EditSummaryFileHappy.nf (hunk 5 above)
Comment (melferink) [reply Utils/EditSummaryFileHappy.nf:14]: makes sense :)

File: CheckQC/check_qc.py (hunk 2 above)
Comment (melferink) [reply CheckQC/check_qc.py:188]: Renamed params for clarification + updated docstring

File: CheckQC/check_qc.py [hunk 6]
```diff
@@ -175,18 +175,48 @@ def create_and_write_output(qc_output, output_path, output_prefix):
...
+            metric_settings["title"],
+            metric_settings["qc_col"],
+            metric_settings["operator"],
+            metric_settings["threshold"]
+        )
+        failed_rows = get_failed_rows(
+            qc_metric_edit,
+            "qc_value",
+            metric_settings["operator"],
+            metric_settings["threshold"]
```
Comment (Jorisvansteenbrugge) [inline CheckQC/check_qc.py:212]: Renaming of the variables and adding the doc strings made this function much clearer!

**Review by Jorisvansteenbrugge — DISMISSED** [review]
LGTM

File: CheckQC/check_qc.py [hunk 7]
```diff
@@ -148,6 +161,14 @@ def add_failed_samples_metric(qc_metric, failed_rows, report_cols, sample_cols):
             drop_index = qc_metric[qc_metric[sample_col].isin(set(failed_samples))].index
             if drop_index.to_list():
                 qc_metric.drop(drop_index, inplace=True)
+
+    # qc_value is stored as string while this should be float.
+    # If ValueError is raised, probably because qc_value is a string (i.e. PASS/FAIL), continue.
```
Comment (Jorisvansteenbrugge) [inline CheckQC/check_qc.py:166]:
```suggestion
    # qc_value can either be a string (i.e., PASS/FAIL), of a floating point value, depending on the QC metric that was used. Float values need to be cast to float types in the dataframe. For string values a ValueError is raised which can be ignored```

**Review by Jorisvansteenbrugge — APPROVED** [review]
kleine suggestie om een comment wat duidelijker/specifieker te maken

File: CheckQC/check_qc.py (hunk 7 above)
Comment (melferink) [reply CheckQC/check_qc.py]: Fair enough

File: CheckQC/check_qc.py [hunk 8]
```diff
@@ -148,6 +161,15 @@ def add_failed_samples_metric(qc_metric, failed_rows, report_cols, sample_cols):
             drop_index = qc_metric[qc_metric[sample_col].isin(set(failed_samples))].index
             if drop_index.to_list():
                 qc_metric.drop(drop_index, inplace=True)
+
+    # qc_value can either be a string (i.e., PASS/FAIL), of a floating point value, depending on the QC metric that was used.
+    # Float values need to be cast to float types in the dataframe to prevent errors in merging of dataframes.
+    # For string values a ValueError is raised which can be ignored
```
Comment (Jorisvansteenbrugge) [inline CheckQC/check_qc.py:167]: Nog beter zo!

**Review by Jorisvansteenbrugge — APPROVED** [review]

**Pull request merged**